optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -235,11 +235,21 @@ private BigDecimalConstraintMatchTotal findBigDecimalConstraintMatchTotal(RuleCo
         return matchTotal;
     }
 
-    private AgendaItem prepareAgendaItemForUnMatchListener(RuleContext kcontext) {
+    private AgendaItem prepareAgendaItemForUnMatchListener(RuleContext kcontext, int scoreLevel) {
         AgendaItem agendaItem = (AgendaItem) kcontext.getMatch();
-        if (agendaItem.getActivationUnMatchListener() != null) {
-            // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
-            agendaItem.getActivationUnMatchListener().unMatch(null, null);
+        ConstraintMatchTotal matchTotal = null;
+        if (constraintMatchEnabled) {
+            Rule rule = kcontext.getRule();
+            matchTotal = constraintMatchTotalMap.get(Arrays.asList(rule.getPackageName(), rule.getName(), scoreLevel));
+            if (agendaItem.getActivationUnMatchListener() != null && matchTotal != null) {
+                // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
+                agendaItem.getActivationUnMatchListener().unMatch(null, null);
+            }
+        } else {
+            if (agendaItem.getActivationUnMatchListener() != null) {
+                // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
+                agendaItem.getActivationUnMatchListener().unMatch(null, null);
+            }
```

Comment:
It doesn't, as I mentioned in the PR description above.

The problem is exactly like you described it in the first comment. Proposal B's solution of a complex ActivationUnMatchListener is probably a better way to go than creating a "unMatchMap" in the score holder, as this would essentially either duplicate or (in the case of constraint matching disabled) imitate the constraintMatchMap. I'll make a quick implementation of B and lets discuss it from there.